File: 2015/06/1.c
#include "input_parser.h"
#include <inttypes.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

struct Lightgrid {
  bool lights[1000][1000];
};
void Lightgrid_new(struct Lightgrid *lightgrid);
void Lightgrid_toggle(struct Lightgrid *lightgrid, const int startx,
                      const int starty, const int endx, const int endy);
void Lightgrid_turnon(struct Lightgrid *lightgrid, const int startx,
                      const int starty, const int endx, const int endy);
void Lightgrid_turnoff(struct Lightgrid *lightgrid, const int startx,
                       const int starty, const int endx, const int endy);
uint32_t Lightgrid_counton(struct Lightgrid *lightgrid);

void Lightgrid_new(struct Lightgrid *lightgrid) {
  for (int x = 0; x < 1000; x++) {
    for (int y = 0; y < 1000; y++) {
      lightgrid->lights[x][y] = false;
    }
  }
}

void Lightgrid_toggle(struct Lightgrid *lightgrid, const int startx,
                      const int starty, const int endx, const int endy) {
  const int stepx = (startx < endx) ? +1 : -1;
  const int stepy = (starty < endy) ? +1 : -1;
  for (int x = startx; x != (endx + stepx); x += stepx) {
    for (int y = starty; y != (endy + stepy); y += stepy) {
      lightgrid->lights[x][y] = !lightgrid->lights[x][y];
    }
  }
}

void Lightgrid_turnon(struct Lightgrid *lightgrid, const int startx,
                      const int starty, const int endx, const int endy) {
  const int stepx = (startx < endx) ? +1 : -1;
  const int stepy = (starty < endy) ? +1 : -1;
  for (int x = startx; x != (endx + stepx); x += stepx) {
    for (int y = starty; y != (endy + stepy); y += stepy) {
      lightgrid->lights[x][y] = true;
    }
  }
}

void Lightgrid_turnoff(struct Lightgrid *lightgrid, const int startx,
                       const int starty, const int endx, const int endy) {
  const int stepx = (startx < endx) ? +1 : -1;
  const int stepy = (starty < endy) ? +1 : -1;
  for (int x = startx; x != (endx + stepx); x += stepx) {
    for (int y = starty; y != (endy + stepy); y += stepy) {
      lightgrid->lights[x][y] = false;
    }
  }
}

uint32_t Lightgrid_counton(struct Lightgrid *lightgrid) {
  uint32_t count = 0;
  for (int x = 0; x < 1000; x++) {
    for (int y = 0; y < 1000; y++) {
      if (lightgrid->lights[x][y]) {
        count++;
      }
    }
  }
  return count;
}
int main() {
  FILE *input = fopen("./2015/06/input.txt", "r");
  if (!input) {
    fprintf(stderr, "Unable to open input file\n");
    return EXIT_FAILURE;
  }
  struct Instruction instruction;
  struct Lightgrid lightgrid;
  Lightgrid_new(&lightgrid);
  while (true) {
    Parse_instruction(input, &instruction);
    switch (instruction.action) {
    case Error:
      fprintf(stderr, "Error parsing input file\n");
      return EXIT_FAILURE;
    case Eof:
      goto end;
    case TurnOn:
      Lightgrid_turnon(&lightgrid, instruction.startx, instruction.starty,
                       instruction.endx, instruction.endy);
      break;
    case TurnOff:
      Lightgrid_turnoff(&lightgrid, instruction.startx, instruction.starty,
                        instruction.endx, instruction.endy);
      break;
    case Toggle:
      Lightgrid_toggle(&lightgrid, instruction.startx, instruction.starty,
                       instruction.endx, instruction.endy);
      break;
    }
  }
end:
  fclose(input);
  printf("%" PRIu32 "\n", Lightgrid_counton(&lightgrid));
  return EXIT_SUCCESS;
}

File: 2015/06/1.go
package main

import (
	"fmt"
	"os"

	"github.com/thunder-coding/advent-of-code/2015/06/input_parser"
)

type Lightgrid struct {
	lights [1000][1000]bool
}

func Lightgrid_turnon(lightgrid *Lightgrid, beg input_parser.Coordinate, end input_parser.Coordinate) {
	var stepx, stepy int
	if beg.X < end.X {
		stepx = +1
	} else {
		stepx = -1
	}
	if beg.Y < end.Y {
		stepy = +1
	} else {
		stepy = -1
	}
	for x := beg.X; x != (end.X + stepx); x += stepx {
		for y := beg.Y; y != (end.Y + stepy); y += stepy {
			lightgrid.lights[x][y] = true
		}
	}
}

func Lightgrid_turnoff(lightgrid *Lightgrid, beg input_parser.Coordinate, end input_parser.Coordinate) {
	var stepx, stepy int
	if beg.X < end.X {
		stepx = +1
	} else {
		stepx = -1
	}
	if beg.Y < end.Y {
		stepy = +1
	} else {
		stepy = -1
	}
	for x := beg.X; x != (end.X + stepx); x += stepx {
		for y := beg.Y; y != (end.Y + stepy); y += stepy {
			lightgrid.lights[x][y] = false
		}
	}
}
func Lightgrid_toggle(lightgrid *Lightgrid, beg input_parser.Coordinate, end input_parser.Coordinate) {
	var stepx, stepy int
	if beg.X < end.X {
		stepx = +1
	} else {
		stepx = -1
	}
	if beg.Y < end.Y {
		stepy = +1
	} else {
		stepy = -1
	}
	for x := beg.X; x != (end.X + stepx); x += stepx {
		for y := beg.Y; y != (end.Y + stepy); y += stepy {
			lightgrid.lights[x][y] = !lightgrid.lights[x][y]
		}
	}
}

func Lightgrid_counton(lightgrid *Lightgrid) uint32 {
	var count uint32
	for x := 0; x != 1000; x++ {
		for y := 0; y != 1000; y++ {
			if lightgrid.lights[x][y] {
				count += 1
			}
		}
	}
	return count
}

func main() {
	input, err := os.Open("./2015/06/input.txt")
	if err != nil {
		fmt.Fprintln(os.Stderr, "Error opening input file")
		os.Exit(1)
	}
	stat, _ := input.Stat()
	size := stat.Size()
	data := make([]byte, size)
	input.Read(data)
	var cur int64 = 0
	var lightgrid Lightgrid
	for {
		instruction, err := input_parser.ParseInstruction(&data, size, &cur)
		if err != nil {
			if err == input_parser.EEof {
				break
			}
			fmt.Fprintf(os.Stderr, err.Error())
			os.Exit(1)
		}

		switch instruction.Action {
		case input_parser.TurnOn:
			Lightgrid_turnon(&lightgrid, instruction.Beg, instruction.End)
		case input_parser.TurnOff:
			Lightgrid_turnoff(&lightgrid, instruction.Beg, instruction.End)
		case input_parser.Toggle:
			Lightgrid_toggle(&lightgrid, instruction.Beg, instruction.End)
		}
	}
	fmt.Println(Lightgrid_counton(&lightgrid))
}
